roachtest: sqlsmith: negative limit hint

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1777546&tab=buildLog) on [master@860c137d602b8420a24be23e097c7fbe1a0a3547](https://github.com/cockroachdb/cockroach/commits/860c137d602b8420a24be23e097c7fbe1a0a3547):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200302-1777546/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:176,sqlsmith.go:206,test_runner.go:741: error: pq: internal error: negative limit hint
		stmt:
		SELECT
			count(*) AS col_40821
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_16857
		GROUP BY
			tab_16857._int8, tab_16857._float8
		HAVING
			similar_to_escape('U_':::STRING::STRING, (WITH with_6841 (col_40822) AS (SELECT * FROM (VALUES (ARRAY['22:03:48.140449':::TIME]), ((ARRAY['23:59:59.999999':::TIME,'00:17:40.351639':::TIME,'13:29:27.637417':::TIME]::TIME[] || '11:04:59.170379':::TIME::TIME)::TIME[]), ((ARRAY['00:34:11.874863':::TIME,'01:09:31.634131':::TIME]::TIME[] || '00:03:14.695':::TIME::TIME)::TIME[]), (ARRAY['18:19:46.447535':::TIME,'23:44:26.263448':::TIME,'23:18:06.63758':::TIME]), (ARRAY[]:::TIME[]), (ARRAY['21:38:34.485071':::TIME,'03:52:52.352767':::TIME])) AS tab_16858 (col_40822)), with_6842 (col_40823) AS (SELECT * FROM (VALUES (()), (())) AS tab_16859 (col_40823)) SELECT substring(e'qXB\x16\x10\\\tW':::STRING::STRING, (json_strip_nulls(json_set(e'[0.6855205127435118, {"W{ dGyAMCyd": {}, "bd^W3?<}": {"Y SGAaAR": []}}, {"c": true}, {"e|p\\"": false, "l,hord2": null}]':::JSONB::JSONB, ((ARRAY[e'\x1eIi\'':::STRING,'':::STRING,e'cwXZQJ\x04':::STRING]::STRING[] || ('mN':::STRING::STRING || NULL::STRING[])::STRING[]::STRING[])::STRING[]::STRING[] || ARRAY[e'z\x01':::STRING]::STRING[])::STRING[]::STRING[], e'[["fY84E&%", "3hW)\\"", [], []], [null, [{}]], {}, {}]':::JSONB::JSONB)::JSONB::JSONB)::JSONB::JSONB#>>ARRAY[e'~\x05':::STRING,e'\x06P':::STRING,'s':::STRING]::STRING[])::STRING::STRING)::STRING AS col_40824 FROM defaultdb.public.seed AS tab_16860 JOIN defaultdb.public.seed AS tab_16861 ON (tab_16860._int8) = (tab_16861._int8) WHERE true GROUP BY tab_16860._date HAVING like_escape(e'~\x15]':::STRING::STRING, e'\x13P{\x12Go\n':::STRING::STRING, e'\b,\x0f[A#l\x17\b':::STRING::STRING)::BOOL ORDER BY tab_16860._date ASC LIMIT 1:::INT8)::STRING, NULL::STRING)::BOOL
		ORDER BY
			tab_16857._float8, tab_16857._int8 ASC, tab_16857._int8 DESC, tab_16857._int8 ASC
		LIMIT
			59:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1777546&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/xform/physical_props.go
+++ b/pkg/sql/opt/xform/physical_props.go
@@ -178,13 +178,17 @@ func BuildChildPhysicalProps(
 // twice as high as needed) tend to occur when distinctCount is very close to
 // neededRows.
 func distinctOnLimitHint(distinctCount, neededRows float64) float64 {
-	if neededRows >= distinctCount {
+	// The harmonic function below is not intended for values under 1 (for one,
+	// it's not monotonic until 0.5); make sure we never return negative results.
+	if neededRows >= distinctCount-1.0 {
 		return 0
 	}
 
 	// Return an approximation of the nth harmonic number.
 	H := func(n float64) float64 {
-		const gamma = 0.5772156649 // Euler–Mascheroni constant
+		// Euler–Mascheroni constant; this is included for clarity but is canceled
+		// out in our formula below.
+		const gamma = 0.5772156649
 		return math.Log(n) + gamma + 1/(2*n)
 	}
 

--- a/pkg/sql/opt/xform/physical_props_test.go
+++ b/pkg/sql/opt/xform/physical_props_test.go
@@ -1,0 +1,47 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package xform
+
+import (
+	"math"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+// TestDistinctOnLimitHint verifies that distinctOnLimitHint never returns a
+// negative result.
+func TestDistinctOnLimitHint(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Generate a list of test values. We want to try the 0 value, very small
+	// values, very large values, and cases where the two values are ~1 off from
+	// each other, or very close to each other.
+	var values []float64
+	for _, v := range []float64{0, 1e-10, 1e-5, 0.1, 1, 2, 3, 10, 1e5, 1e10} {
+		values = append(values, v)
+		for _, noise := range []float64{1e-10, 1e-7, 0.1, 1} {
+			values = append(values, v+noise)
+			if v-noise >= 0 {
+				values = append(values, v-noise)
+			}
+		}
+	}
+
+	for _, distinctCount := range values {
+		for _, neededRows := range values {
+			hint := distinctOnLimitHint(distinctCount, neededRows)
+			if hint < 0 || math.IsNaN(hint) || math.IsInf(hint, +1) {
+				t.Fatalf("distinctOnLimitHint(%g,%g)=%g", distinctCount, neededRows, hint)
+			}
+		}
+	}
+}

--- a/pkg/sql/opt/xform/testdata/physprops/limit_hint
+++ b/pkg/sql/opt/xform/testdata/physprops/limit_hint
@@ -242,21 +242,22 @@ limit
 
 # DistinctOn operator.
 opt
-SELECT DISTINCT z FROM t LIMIT 2
+SELECT DISTINCT z FROM t LIMIT 1
 ----
 limit
  ├── columns: z:3
- ├── cardinality: [0 - 2]
- ├── key: (3)
+ ├── cardinality: [0 - 1]
+ ├── key: ()
+ ├── fd: ()-->(3)
  ├── distinct-on
  │    ├── columns: z:3
  │    ├── grouping columns: z:3
  │    ├── key: (3)
- │    ├── limit hint: 2.00
+ │    ├── limit hint: 1.00
  │    └── scan t
  │         ├── columns: z:3
- │         └── limit hint: 3.44
- └── 2
+ │         └── limit hint: 1.23
+ └── 1
 
 # No limit hint propagation if number of distinct rows < required number of rows.
 opt

